trio/_util.py

```diff
@@ -138,3 +147,44 @@ def helper(*args, **kwds):
     # A hint for sphinxcontrib-trio:
     helper.__returns_acontextmanager__ = True
     return helper
+
+
+class _UnLockSync:
+    def __init__(self, exc, *args):
+        self._exc = exc
+        self._args = args
+        self._held = False
+
+    def __enter__(self):
+        if self._held:
+            raise self._exc(*self._args)
+        else:
+            self._held = True
+
+    def __exit__(self, *args):
+        self._held = False
+
+
+class UnLock:
```

Comment:
Yeah, this is definitely a bit over-cutesy. (Note that it isn't a public API, though.) Also it should probably be refactored a bit since currently it's written as a generic object that can raise any exception, but in fact every single caller uses it to raise `ResourceBusy`.